tpm.go

```diff
@@ -0,0 +1,937 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	_ "crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/chrisccoulson/go-tpm2"
+	"github.com/snapcore/snapd/httputil"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/sys"
+
+	"golang.org/x/xerrors"
+)
+
+const (
+	// FIXME: This is fine during initial install and early boot, but we should strive to use the resource manager at other times.
+	tpmPath string = "/dev/tpm0"
+
+	// Handle for RSA2048 EK certificate, see section 7.8 of "TCG TPM v2.0 Provisioning Guidance" Version 1.0, Revision 1.0, 15 March 2017.
+	ekCertHandle tpm2.Handle = 0x01c00002
+
+	sanDirectoryNameTag = 4 // Subject Alternative Name directoryName, see section 4.2.16 or RFC5280
+)
+
+var (
+	oidExtensionSubjectAltName = asn1.ObjectIdentifier{2, 5, 29, 17} // id-ce-subjectAltName, see section 4.2.16 of RFC5280
+
+	// TCG specific OIDs, see section 4 of "TCG EK Credential Profile For TPM Family 2.0; Level 0", Version 2.1, Revision 13, 10 December 2018.
+	oidTcgAttributeTpmManufacturer = asn1.ObjectIdentifier{2, 23, 133, 2, 1} // tcg-at-tpmManufacturer
+	oidTcgAttributeTpmModel        = asn1.ObjectIdentifier{2, 23, 133, 2, 2} // tcg-at-tpmModel
+	oidTcgAttributeTpmVersion      = asn1.ObjectIdentifier{2, 23, 133, 2, 3} // tcg-at-tpmVersion
+	oidTcgKpEkCertificate          = asn1.ObjectIdentifier{2, 23, 133, 8, 1} // tcg-kp-EKCertificate
+)
+
+// TPMDeviceAttributes contains details about the TPM extracted from a manufacturer issued endorsement key certificate.
+type TPMDeviceAttributes struct {
+	Manufacturer    tpm2.TPMManufacturer
+	Model           string
+	FirmwareVersion uint32
+}
+
+// TPMConnection corresponds to a connection to a TPM device, and is a wrapper around *tpm2.TPMContext.
+type TPMConnection struct {
+	*tpm2.TPMContext
```

Comment:
will caller use methods from this directly?

Reply:
Yes, possibly - specifying the authorization value for permanent handles requires it (eg, see https://godoc.org/github.com/chrisccoulson/go-tpm2#TPMContext.OwnerHandleContext and https://godoc.org/github.com/chrisccoulson/go-tpm2#ResourceContext). Also, it's possible that some errors could require corrective action, such as undefining a NV index (using tpm2.TPMContext.NVUndefineSpace).